### PR TITLE
Replace scap_strerror_* with scap_errprintf

### PR DESCRIPTION
**Pull request by gnosek**
`scap_strerror()` has a fairly clunky API: it requires the full`scap_t` handle, even though it only accesses `m_strerror_buf`. `scap_strerror_r` (introduced recently) relaxes this requirement but still I believe the error reporting UX isn't optimal (we often need a temporary buffer just to format the errno message).

Looking at the call sites, it seems to me that a printf-like function that appends the error message automatically handles most common use cases. I also made it return `SCAP_FAILURE`, so the typical usage would be:

```
if(try_something(x) < 0)
{
    return scap_errprintf(handle->m_lasterr, errno, "tried something with %d and it failed", x);
}
```

which would get e.g. `: Permission denied` appended if errno == EPERM.

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

/area libscap-engine-bpf

> /area libscap-engine-gvisor

/area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

/area libscap-engine-udig

/area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

Even though this changes the libscap API, I don't consider it a user-facing change, since `scap_getlasterr` remains the same and scap users shouldn't generate their own scap errors IMO. This is relevant mostly (only?) for engine developers.

Marked as `wip` since I realized I haven't audited all the raw bpf syscalls correctly (the error reporting was almost certainly broken before too but since I'm already messing with this area, now's a good time to clean this up).

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by gnosek** [comment]
I just added fixed a Windows build error a new commit which isn't really related (Windows udig cleanups/fixes) so if you review it, please take a separate look at it. I can also submit it as a separate PR if preferred but honestly I'd rather keep it on top of this one due to conflicts.

So, I can:
* leave it as is
* submit a new PR after this one lands
* submit an independent PR and fix conflicts in whatever is not merged first

Thoughts?

File: userspace/libscap/ringbuffer/devset.c [hunk 1]
```diff
@@ -28,11 +21,11 @@ int32_t devset_init(struct scap_device_set *devset, size_t num_devs, char *laste
...
-		devset->m_devs[j].m_buffer = MAP_FAILED;
-		devset->m_devs[j].m_bufinfo = MAP_FAILED;
-		devset->m_devs[j].m_bufstatus = MAP_FAILED;
-		devset->m_devs[j].m_fd = -1;
-		devset->m_devs[j].m_bufinfo_fd = -1;
+		devset->m_devs[j].m_buffer = INVALID_MAPPING;
+		devset->m_devs[j].m_bufinfo = INVALID_MAPPING;
+		devset->m_devs[j].m_bufstatus = INVALID_MAPPING;
+		devset->m_devs[j].m_fd = INVALID_FD;
+		devset->m_devs[j].m_bufinfo_fd = INVALID_FD;
```
Comment (Molter73) [inline userspace/libscap/ringbuffer/devset.c:28]: I don't understand why some of these refactorings of devset handling are part of this PR 🤔. Don't get me wrong, they are awesome and greatly simplify the code, I just wish they were in their own PR so they could get a bit more of the spotlight.

**Review by Molter73 — APPROVED** [review]
/lgtm

Thanks for this change @gnosek!

File: userspace/libscap/ringbuffer/devset.c (hunk 1 above)
Comment (gnosek) [reply userspace/libscap/ringbuffer/devset.c:28]: Yeah, that's what my last comment is about :) I can submit it separately after just the scap_strerror cleanup if you prefer (I'd just rather not do another series of rebase dances and it depends a bit on the scap_strerror cleanup).

It does fix rather important Windows warnings (how in the world didn't we get a link error?) so I'd rather see it in sooner than later but I'm fine with a separate PR.

**Comment by gnosek** [comment]
@Molter73, removed the last commit. I'll submit it again once this gets merged.

**Review by Molter73 — APPROVED** [review]
/lgtm

**Review by FedeDP — COMMENTED** [review]
I've yet to finish an in-depth reviews; the following were a couple of things i quickly noticed.

File: userspace/libscap/win32/strerror.c [hunk 2]
```diff
@@ -17,30 +17,30 @@ limitations under the License.
...
 #include <stdio.h>
 #include <string.h>
 
 #include "scap.h"
 #include "scap-int.h"
 
-#define strerror_r(errnum, buf, size) strerror_s(buf, size, errnum)
-
-const char *scap_strerror_r(char *buf, int errnum)
+int32_t scap_errprintf_unchecked(char *buf, int errnum, const char* fmt, ...)
```
[the remark this answers is not in the document]
Comment (gnosek) [reply userspace/libscap/win32/strerror.c:27]: strerror_r vs strerror_s. I guess we can unify them back with an #ifdef. Thoughts?

File: userspace/libscap/engine/bpf/scap_bpf.c [hunk 3]
```diff
@@ -398,9 +397,7 @@ static int32_t load_maps(struct bpf_engine *handle, struct bpf_map_data *maps, i
 
 		if(handle->m_bpf_map_fds[j] < 0)
 		{
-			char buf[SCAP_LASTERR_SIZE];
-			snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "can't create map: %s", scap_strerror_r(buf, errno));
-			return SCAP_FAILURE;
+			return scap_errprintf(handle->m_lasterr, -handle->m_bpf_map_fds[j], "can't create map %d", j);
```
[the remark this answers is not in the document]
Comment (gnosek) [reply userspace/libscap/engine/bpf/scap_bpf.c:400]: Yes and it was broken. Raw syscalls don't set errno.

File: userspace/libscap/win32/strerror.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (FedeDP) [reply userspace/libscap/win32/strerror.c:27]: I think so :)

**Comment by Andreagit97** [comment]
@gnosek could you rebase when you have time :)?

**Comment by gnosek** [comment]
Rebased and unified posix/win32 implementations. This required introducing a new (tiny) library for that single file, otherwise we'd have cyclic dependencies somewhere, or link a file from userspace/libscap into scap_platform, neither of which seemed desirable.

**Comment by FedeDP** [comment]
Perhaps in the future we will add more common functions to this new library. It's ok for me. Thank you very much!

**Review by FedeDP — APPROVED** [review]
/approve

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**